Support Tahoe 2.0 site configuration client

OmarIthawi opened the pull request:
Fix [RED-3385](https://appsembler.atlassian.net/browse/RED-3385). Read Tahoe 2 site from `secret` on Tahoe 2.0 sites and from `site_values` on Tahoe 1.0 sites.

### Breaking change
Tahoe Hawthorn is no longer supported in v4.0.0 due to the lack of [Site Configuration Client](https://github.com/appsembler/site-configuration-client/). It will act as if it's installed on a Vanilla Hawthorn.

### TODO
 - [x] Test on devstack with site configuration on a Tahoe 2.0 site
   - [x] with site-specific value
   - [x] with `ENV_TOKENS` value
 - [x] Test on devstack with site configuration on a Tahoe 1.0 site
   - [x] with site-specific value
   - [x] with `ENV_TOKENS` value

briandant commented on launchcontainer/test_settings.py:
```diff
@@ -0,0 +1,29 @@
+"""
+Test settings for the Feedback XBlock.
```
```suggestion
Test settings for the Launchcontainer XBlock.
```

briandant commented on launchcontainer/tests.py:
```diff
@@ -128,10 +141,27 @@ def test_student_view(self, fragment, render_template):
-    @mock.patch('launchcontainer.launchcontainer.load_resource', DummyResource)
+    def test_author_preview_view(self):
+        """
+        Test author preview view renders correctly with and without .
```
```suggestion
        Test author preview view renders correctly with and without a wharf endpoint.
```

OmarIthawi committed:
changelog for v4.0.0

OmarIthawi committed:
Support Tahoe 2.0 site configuration client

 - fix author_view when wharf_url is missing
 - support rendering templates in tests
 - new author_view for studio
 - refactored tests
 - Tahoe Hawthorn is no longer supported in v4.0.0 due to the lack of Site Configuration Client. It will act as if it's installed on a Vanilla Hawthorn.
 - Install `tahoe-sites` and `site-configuration-client` to simulate Tahoe environment.
 - Add support for Tahoe 2.0 site by reading `LAUNCHCONTAINER_WHARF_URL` from Site Config Service `secret`s. Tahoe 1.0 sites will still use `site_values` as usual
 - Breaking change: Drop support for pre-Juniper Tahoe installations. The last version to support Hawthorn is `3.0.0`
need to complicate the XBlock with caching calls.

 - Simplify reading the Wharf URL
   * Read from `ENV_TOKENS` for regular Open edX installation (non-Tahoe)
   * Read from [Site Configuration Client](https://github.com/appsembler/site-configuration-client/) for Juniper+ installations

 - Refactored unit tests
 - Refactor testing environments in `tox.ini` to test for `tahoe` and non `tahoe` environments removing `tox-gh-actions`.
 - Remove complex and low-value caching for `wharf_url`

briandant commented on CHANGELOG.md:
```diff
@@ -6,8 +6,24 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+ - Install `tahoe-sites` and `site-configuration-client` to simulate Tahoe environment.
+ - Add support for Tahoe 2.0 site by reading `LAUNCHCONTAINER_WHARF_URL` from Site Config Service `secret`s. Tahoe 1.0 sites will still use `site_values` as usual
+- Breaking change: Drop support for pre-Juniper Tahoe installations. The last version to support Hawthorn is `3.0.0`
+need to complicate the XBlock with caching calls.
```
```suggestion
- Do not complicate the XBlock with caching calls.
```

briandant replied:
actually I think you can delete this line—it's duplicated below

briandant commented on CHANGELOG.md:
```diff
@@ -6,8 +6,24 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+
+ - Install `tahoe-sites` and `site-configuration-client` to simulate Tahoe environment.
+ - Add support for Tahoe 2.0 site by reading `LAUNCHCONTAINER_WHARF_URL` from Site Config Service `secret`s. Tahoe 1.0 sites will still use `site_values` as usual
+- Breaking change: Drop support for pre-Juniper Tahoe installations. The last version to support Hawthorn is `3.0.0`
```
```suggestion
 - Breaking change: Drop support for pre-Juniper Tahoe installations. The last version to support Hawthorn is `3.0.0`
```

briandant commented on launchcontainer/launchcontainer.py:
```diff
@@ -143,82 +171,36 @@ class LaunchContainerXBlock(XBlock, xblocksettings.XBlockWithSettingsMixin):
 
-        return url
+    def get_wharf_delete_url(self, required=True):
+        wharf_url = self.get_wharf_url(required=required)
```
I'm not sure I totally understand why we check for a wharf url and then return None.  I *think* it's because we want this method (`get_wharf_delete_url`) to also make it clear that no URL is available.  I'm also not sure why we'd return a delete URL if `required=False` but there is no `wharf_url`.  I just need a docstring or comment here, I think.

briandant commented on launchcontainer/static/html/launchcontainer.html:
```diff
@@ -8,5 +9,10 @@
   <div id="launcher_notification" class="hide"></div>
   {% if enable_container_resetting %}<button type="button" class="ui-priority-primary" id="launcher_reset">Reset lab</button>{% endif %}
-
+{% else %}
```
Great addition here 👍

briandant commented on launchcontainer/static/html/launchcontainer.html:
```diff
@@ -8,5 +9,10 @@
+{% else %}
+  {# This message is intended to appear Studio for course authors. #}
+  <p><strong>Error: Unable to connect to the Appsembler Virtual Labs API endpoint.</strong></p>
+  <p>Please configure your Appsembler Virtual Labs API endpoint before using this XBlock.</p>
```
```suggestion
  <p>Please contact  <a href="mailto:support@appsembler.com" target="_blank">Appsembler Support</a> so we can configure your Appsembler Virtual Labs API endpoint.</p>
```

I could be wrong, but I don't think that the customer can configure this on his own, right?  My suggested phrasing above captures that.  If he can, I think we should talk with Matthew about writing some customer docs on that (or maybe the docs already exist).

briandant reviewed (approved):
many thanks, @OmarIthawi!

OmarIthawi commented:
Thanks Brian!! I'll follow up on all additions in another PR to avoid GitHub invalidating your review.
